README.md
# logtoools


debian/changelog
logtools (0.13d) unstable; urgency=low

  * Merged changes from NMU and fixed some more GCC 4.3 warnings.

 -- Russell Coker <russell@coker.com.au>  Fri, 06 Jun 2008 15:34:35 +1000

logtools (0.13c+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with GCC 4.3 (Closes: #417376).

 -- Luk Claes <luk@debian.org>  Sun, 09 Mar 2008 17:03:20 +0000

logtools (0.13c) unstable; urgency=low

  * Made it compile with GCC 3.2.

 -- Russell Coker <russell@coker.com.au>  Fri, 25 Apr 2003 17:09:00 +1000

logtools (0.13b) unstable; urgency=low

  * Improved documentation that matches how things work in sarge.

 -- Russell Coker <russell@coker.com.au>  Sun,  1 Sep 2002 15:43:00 +0200

logtools (0.13a) unstable; urgency=LOW

  * Made clfdomainsplit support URLs with the port number.

 -- Russell Coker <russell@coker.com.au>  Sun,  2 Dec 2001 23:58:00 +0100

logtools (0.13) unstable; urgency=LOW

  * Made it work with GCC 3.x.
    Closes: #121284

 -- Russell Coker <russell@coker.com.au>  Mon, 26 Nov 2001 21:46:00 +0100

logtools (0.12b) unstable; urgency=LOW

  * Fixed some minor documentation bugs, mainly related to funnel.
    Closes: #112883

 -- Russell Coker <russell@coker.com.au>  Thu, 20 Sep 2001 15:40:02 +0200

logtools (0.12a) unstable; urgency=LOW

  * Fixed header file errors on funnel for hppa.
    Closes: #104730

 -- Russell Coker <russell@coker.com.au>  Sat, 14 Jul 2001 20:04:55 +0200

logtools (0.12) unstable; urgency=LOW

  * clfmerge -d now converts domain names to lower case.

  * clfdomainsplit now converts all domain names to lower case internally
    (for sorting and matching purposes), it still exposes upper-case domains
    on its output.

 -- Russell Coker <russell@coker.com.au>  Wed, 11 Jul 2001 17:09:00 +0200

logtools (0.11) unstable; urgency=LOW

  * Made the "-d" option of clfmerge handle a URL that doesn't contain any '/'
    characters.

 -- Russell Coker <russell@coker.com.au>  Tue, 12 Jun 2001 11:58:49 +0200

logtools (0.10) unstable; urgency=MEDIUM

  * Made clfdomainsplit handle IP addresses and send them to the default file
    (first code to actually use the default file).

  * Fixed clfmerge so that the -d option will not mess with a URL that is
    already fully qualified (checks for "http://" at the start).

 -- Russell Coker <russell@coker.com.au>  Thu, 7 Jun 2001 23:49:55 +0200

logtools (0.08) unstable; urgency=MEDIUM

  * Made clfdomainsplit easier to hack (the base code could be used for
    splitting any text file if 3 small functions are changed).

  * Make clfdomainsplit use multiple processes to solve file handle limits,
    this code isn't yet tested...

 -- Russell Coker <russell@coker.com.au>  Mon, 8 May 2001 13:50:29 +0200

logtools (0.07) unstable; urgency=MEDIUM

  * Updated the description with more content and fixed some formatting issues.
    Closes: #78976

  * Made it clean the config.cache file (and other config data) before
    packaging the source.
    Closes: #96290

  * Fixed a bug where clfmerge would output the last line twice if there was
    only a single log file and buffering was on.

  * Now "domain name mangling" actually works for clfmerge!

  * Added new clfdomainsplit program.

 -- Russell Coker <russell@coker.com.au>  Mon, 7 May 2001 13:52:41 +0200

logtools (0.06) unstable; urgency=low

  * Add option to change buffer size for clfmerge.

  * Also added an option for "domain name mangling" to clfmerge.  This means
    that the web logs are expected to have a domain name preceeding the regular
    CLF format (as suggested in the Apache documentation for mass virtual
    hosting).
    See http://www.apache.org/DELETE_ME/vhosts/mass.html .

  * Fixed a bunch of errors in the man pages.

 -- Russell Coker <russell@coker.com.au>  Wed, 2 May 2001 18:11:24 +0200

logtools (0.05) unstable; urgency=low

  * Fixed many trivial bugs.

 -- Russell Coker <russell@coker.com.au>  Tue, 5 Dec 2000 21:28:54 +0100

logtools (0.04) unstable; urgency=low

  * Fixed heaps of bugs and ported to Solaris.

 -- Russell Coker <russell@coker.com.au>  Tue, 14 Nov 2000 21:39:00 +0100

logtools (0.01) unstable; urgency=low

  * Initial Release.

 -- Russell Coker <russell@coker.com.au>  Fri, 25 Aug 2000 19:53:08 +0200



funnel.cpp
#include <unistd.h>
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>

#include "logtools.h"

void usage()
{
  fprintf(stderr, "Usage: funnel [|] [>[>]file] [|program]\n"
  "\nVersion: " VERSION "\n");
  exit(ERR_PARAM);
}

PFILE openit(const char *arg)
{
  const char *ptr = arg;
  PFILE ret = NULL;

  if(!strcmp(arg, "|"))
    return stdout;

  if(ptr[0] == '>')
  {
    const char *mode = "w";
    ptr++;
    if(ptr[0] == '>')
    {
      ptr++;
      mode = "a";
    }
    ret = fopen(ptr, mode);
  }
  else if(ptr[0] == '|')
  {
    ptr++;
    ret = popen(ptr, "w");
  }
  else
  {
    fprintf(stderr, "Bad command \"%s\".\n", arg);
    return NULL;
  }
  return ret;
}

int main(int argc, char **argv)
{
  if(argc < 2)
    usage();

  PFILE *out = new PFILE[argc];

  int i;
  for(i = 1; i < argc; i++)
  {
    out[i] = openit(argv[i]);
  }
  char buf[8192];
  size_t rc;
  int count = argc - 1;
  int errors = 0;
  int f = fcntl(0, F_GETFL);
  if(f == -1 || -1 == fcntl(0, F_SETFL, f & (!O_NONBLOCK)) )
  {
    fprintf(stderr, "Can't fcntl\n");
    return ERR_INPUT;
  }

  while( int(rc = read(0, buf, sizeof(buf)) ) > 0 && count > 0)
  {
    if(rc)
    {
      for(i = 1; i < argc; i++)
      {
        if(out[i])
        {
          size_t wrote = fwrite(buf, rc, 1, out[i]);
          if(wrote != 1)
          {
            fclose(out[i]);
            out[i] = NULL;
            fprintf(stderr, "Output \"%s\" aborted.\n", argv[i]);
            count--;
            errors++;
          }
        }
      }
      fflush(NULL);
    }
  }
  close(0);
  for(i = 1; i < argc; i++)
  {
    if(out[i])
      fclose(out[i]);
  }
  return 100 + errors;
}



logprn.cpp
#include <stdio.h>
#include <sys/stat.h>
#include <unistd.h>
#include <time.h>
#include <dirent.h>
#include <stdlib.h>
#include <sys/types.h>
#include <fcntl.h>
#include <string.h>

#include "logtools.h"

void usage()
{
  fprintf(stderr, "Usage: logprn logfile idle-time[:max-wait] command\n"
  "\nVersion: " VERSION "\n");
  exit(ERR_PARAM);
}

int main(int argc, char **argv)
{
  if(argc != 4)
    usage();

  struct stat sBuf;
  int rc = stat(argv[1], &sBuf);
  if(rc)
  {
    printf("Can't stat \"%s\".\n", argv[1]);
    return ERR_PARAM;
  }

  bool changed = false;
  time_t mtime = sBuf.st_mtime;
  ino_t inode = sBuf.st_ino;
  off_t size = sBuf.st_size;
  off_t print_size = size;

  char *pbuf = strdup(argv[2]);
  strtok(pbuf, ":");
  char *maxBuf = strtok(NULL, ":");
  time_t delay = atoi(pbuf);
  time_t maxWait = 0;
  if(maxBuf)
    maxWait = atoi(maxBuf);
  free(pbuf);
  if(maxWait && maxWait < delay)
    usage();

  if(delay < 1)
    usage();
  time_t last_change = time(NULL);
  time_t first_unwritten_change = 0;

  while(1)
  {
    sleep(1);
    rc = stat(argv[1], &sBuf);
    if(rc)
    {
      // if failed then try again 1 second later in case of link changes etc.
      sleep(1);
      rc = stat(argv[1], &sBuf);
    }
    if(rc)
    {
      fprintf(stderr, "File disappeared or became unreadable.\n");
      return ERR_INPUT;
    }
    if(inode != sBuf.st_ino)
    {
      inode = sBuf.st_ino;
      changed = true;
      mtime = sBuf.st_mtime;
      print_size = 0;
      size = sBuf.st_size;
      last_change = time(NULL);
    }
    else if(mtime != sBuf.st_mtime || size != sBuf.st_size)
    {
      if(size > sBuf.st_size)
        print_size = 0;
      size = sBuf.st_size;
      mtime = sBuf.st_mtime;
      changed = true;
      last_change = time(NULL);
      if(first_unwritten_change == 0)
        first_unwritten_change = last_change;
    }
    time_t now = time(NULL);
    if(changed)
    {
      if((now - last_change) > delay
       || (maxWait && (now - first_unwritten_change ) > maxWait) )
      {
        int fd = open(argv[1], O_RDONLY);
        if(fd == -1)
        {
          fprintf(stderr, "Can't open file \"%s\"\n", argv[1]);
          return ERR_INPUT;
        }
        rc = lseek(fd, print_size, SEEK_SET);
        if(rc == -1)
        {
          fprintf(stderr, "Can't lseek().\n");
        }
        FILE *fp = popen(argv[3], "w");
        if(!fp)
        {
          fprintf(stderr, "Can't run \"%s\"\n", argv[3]);
          return ERR_OUTPUT;
        }
        char buf[4096];
        while( (rc = read(fd, buf, sizeof(buf))) > 0)
        {
          if(int(fwrite(buf, 1, rc, fp)) != rc)
          {
            fprintf(stderr, "Short write to pipe.\n");
            break;
          }
          print_size += rc;
        }

        pclose(fp);
        close(fd);
        changed = false;
        first_unwritten_change = 0;
      }
    }
  }
  return 0; // to make gcc happy
}



clfsplit.cpp
#include <cstdlib>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <map>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include "logtools.h"

using namespace std;

void usage()
{
  fprintf(stderr
        , "Usage: clfsplit -d defaultfile -f file -s spec [-f file -s spec]\n"
          "\nSpec is a list of IP ranges or a file containing such a list.\n"
  "\nVersion: " VERSION "\n");
  exit(ERR_PARAM);
}

class file_wrap
{
public:
  file_wrap(FILE *fp, const char *name)
   : m_fp(fp)
   , m_name(name)
  { }


  void close() { if(m_fp) { fclose(m_fp); m_fp = NULL; } }

  void write(const char *line)
  {
    if(EOF == fputs(line, m_fp))
    {
      fprintf(stderr, "Can't write to output file %s\n", m_name);
      exit(ERR_OUTPUT);
    }
  }

private:
  FILE *m_fp;
  const char *m_name;

  file_wrap(const file_wrap&);
  file_wrap & operator=(const file_wrap&);
};

class details;
typedef map<int, details *> MAP;

const char *inet_itoa(unsigned int ip)
{
  static char buf[30];
  sprintf(buf, "%d.%d.%d.%d", ip >> 24, (ip >> 16) % 256, (ip >> 8) % 256, ip % 256);
  return buf;
}

class details
{
public:
  details(unsigned int begin, unsigned int end, file_wrap *wrap, MAP &m)
   : m_begin(begin)
   , m_end(end)
   , m_wrap(wrap)
  {
    m[begin] = this;
  }

  ~details() { m_wrap->close(); }

  bool check_item(unsigned int ip, const char *line)
  {
    if(ip >= m_begin && ip <= m_end)
    {
      m_wrap->write(line);
      return true;
    }
    return false;
  }

private:
  unsigned int m_begin, m_end;
  file_wrap *m_wrap;

  details(const details&);
  details & operator=(const details&);
};


unsigned int inet_atoi(const char *n)
{
  char tmp[16];
  strncpy(tmp, n, sizeof(tmp) - 1);
  tmp[sizeof(tmp) - 1] = '\0';
  for(unsigned int i = 0; i < sizeof(tmp); i++)
  {
    if(tmp[i] == ' ')
      tmp[i] = '\0';
  }
  return ntohl(inet_addr(tmp));
}

FILE *open_output_file(const char *file)
{
  if(!file)
    usage();

  const char *mode = "w";
  struct stat stat_buf;
  if(!stat(file, &stat_buf) && S_ISREG(stat_buf.st_mode))
    mode = "a";
  FILE *output = fopen(file, mode);
  if(!output)
    fprintf(stderr, "Can't open file %s\n", file);
  return output;
}

void process_spec(MAP &m, const char *file, const char *opt)
{
  FILE *output = open_output_file(file);
  if(!output)
    exit(ERR_OUTPUT);
  file_wrap *wrap = new file_wrap(output, file);

  FILE *fp = NULL;
  char buf[4096];
  char *ptr = buf;
  if(opt[0] < '0' || opt[0] > '9')
  {
    fp = fopen(opt, "r");
    if(!fp)
    {
      fprintf(stderr, "Can't open file %s\n", opt);
      exit(ERR_SPEC);
    }
    if(!fgets(buf, sizeof(buf), fp) )
    {
      fprintf(stderr, "Can't read from file %s\n", opt);
      exit(ERR_SPEC);
    }
  }
  else
  {
    ptr = strdup(opt);
  }
  do
  {
    buf[sizeof(buf) - 1] = '\0';

    char *item;
    item = strtok(ptr, "\r\n:");
    do
    {
      int i;
      char *it = strdup(item);
      for(i = 0; item[i] != '\0' && item[i] != '-' && item[i] != '/'; i++)
      { }

      unsigned int first_ip = 0;
      unsigned int second_ip = 0;
      switch(item[i])
      {
        case '\0':
          first_ip = inet_atoi(item);
          second_ip = first_ip;
        break;
        case '-':
          item[i] = '\0';
          first_ip = inet_atoi(item);
          second_ip = inet_atoi(&item[i + 1]);
        break;
        case '/':
        {
          item[i] = '\0';
          first_ip = inet_atoi(item);
          unsigned int sub = atoi(&item[i + 1]);
          second_ip = first_ip + (1 << (32 - sub)) - 1;
        }
        break;
      }
      new details(first_ip, second_ip, wrap, m);
      free(it);
    } while( (item = strtok(NULL, "\r\n:")) );

  } while(fp && fgets(buf, sizeof(buf), fp) );
  if(fp)
    fclose(fp);
  else
    free(ptr);
}


int main(int argc, char **argv)
{
  int int_c;
  const char *defaultfile = NULL;
  const char *file = NULL;
  MAP m;
  FILE *input = stdin;
  bool new_input = false;
  optind = 0;
  while(-1 != (int_c = getopt(argc, argv, "d:i:f:s:")) )
  {
    switch(char(int_c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'd':
        defaultfile = optarg;
      break;
      case 'i':
        if(new_input)
          usage();
        input = fopen(optarg, "r");
        if(!input)
        {
          fprintf(stderr, "Can't open file %s\n", optarg);
          return ERR_INPUT;
        }
        new_input = true;
      break;
      case 'f':
        file = optarg;
      break;
      case 's':
        process_spec(m, file, optarg);
      break;
    }
  }
  new details(0, 0, NULL, m);
  FILE *def = open_output_file(defaultfile);
  if(!def)
    return ERR_OUTPUT;
  char buf[4096];
  while(fgets(buf, sizeof(buf), input))
  {
    buf[sizeof(buf) - 1] = '\0';
    unsigned int ip = inet_atoi(buf);
    if(ip)
    {
      MAP::iterator i = m.upper_bound(ip);
      bool done = false;
      if(i != m.begin() )
      {
        i--;
        done = (*i).second->check_item(ip, buf);
      }
      if(!done)
      {
        if(EOF == fputs(buf, def))
        {
          fprintf(stderr, "Can't write to output file %s\n", defaultfile);
          return ERR_OUTPUT;
        }
      }
    }
  }
  if(new_input)
    fclose(input);
  return 0;
}



clfdomainsplit.cpp
#include <stdio.h>

#if (__GNUC__ >= 3)
#include <ext/hash_map>
#else
#include <hash_map>
#endif

#include <unistd.h>
#include <cstring>
#include <vector>
#include <ctype.h>
#include "logtools.h"

using namespace std;
using namespace __gnu_cxx;

// MAX_FDS is the maximum number of files that will be directly written to
// by one process
#define MAX_FDS  900

// NUM_CHILDREN is the number of child processes that each process may have
// Note that each child may also have that many children.  The limit is the
// size of the process table in your kernel...
#define NUM_CHILDREN  100

// Get the "domain" component of the line.  Change this if you want to split
// on other parts of the name (EG split files on the directory names).  Make
// sure that the value assigned to dom is a valid file name.
// Change this if you want to use this program for splitting files other than
// web logs.
bool getDomain(string &dom, PCCHAR buf)
{
  if(!(buf = index(buf, '[')) ) return true;
  if(!(buf = index(buf, ']')) ) return true;
  if(!(buf = index(buf, '"')) ) return true;
  if(!(buf = strstr(buf, "http://")) ) return true;
  buf += 7;

  PCCHAR end = index(buf, '/');
  if(!end) return true;
  PCCHAR tmp_end = index(buf, ':'); //search for ":80/" at the end and ignore it
  if(tmp_end && tmp_end < end)
    end = tmp_end;
  int len = end - buf;
  PCHAR tmp = new char[len];
  memcpy(tmp, buf, len);
  for(int i = 0; i < len; i++)
    tmp[i] = tolower(tmp[i]);
  dom = string(tmp, int(len));
  free(tmp);
  return false;
}

// The data structure for configuration.  If you change the getDomain() and
// canonDom() functions then you'll probably need to change this and the
// parseConfig() function.
typedef struct
{
  string domain;
  int significant_parts;
} CFG_TYPE;
vector<CFG_TYPE> config;

// Get the canonical name
bool canonDom(string &canon, string &dom)
{
  int last = dom.length() - 1;
  canon = dom;
  if(canon[last] == '.')
  {
    canon.erase(last);
    last--; // for names that end in '.'
  }

  if(isdigit(canon[last]))
    return true; // canon == dom for numeric ID, and it's not valid for split

  int significant_parts = 2;
  unsigned int i;
  for(i = 0; i < config.size(); i++)
  {
    size_t pos = dom.rfind(config[i].domain);
    if(int(pos) != -1 && (pos + config[i].domain.length()) == dom.length())
    {
      significant_parts = config[i].significant_parts;
      break;
    }
  }
  for(i = last; i > 0; i--)
  {
    if(dom[i] == '.')
    {
      significant_parts--;
      if(!significant_parts)
      {
        canon.assign(dom, i + 1, dom.length() - i - 1);
        break;
      }
    }
  }
  return false;
}

void usage();

// This function parses the config file.
bool parseConfig(CPCCHAR configFile)
{
  FILE *fp = fopen(configFile, "r");
  if(!fp)
  {
    fprintf(stderr, "Can't open \"%s\".\n", configFile);
    usage();
  }
  char buf[1024];
  CFG_TYPE cfg;
  while(fgets(buf, sizeof(buf), fp))
  {
    buf[sizeof(buf) - 1] = '\0';
    strtok(buf, "\r\n");
    if(strlen(buf) && buf[0] != '#')
    {
      cfg.significant_parts = atoi(buf);
      if(cfg.significant_parts <= 0 || cfg.significant_parts > 10)
      {
        fprintf(stderr, "Error in config line \"%s\".\n", buf);
        return true;
      }
      char *ptr = index(buf, ':');
      ptr++;
      if(*ptr == '\0')
      {
        fprintf(stderr, "Error in config line \"%s\".\n", buf);
        return true;
      }
      cfg.domain = string(ptr);
      config.push_back(cfg);
    }
  }
  fclose(fp);
  return false;
}


FILE **children[NUM_CHILDREN];
int num_fds = 0;
int newTarget = -1; // -1 means handle locally, anything else means the
                    // current child process target
FILE *input = stdin;
PFILE *defFilePtr;
int num_children = 0;



struct eqstr
{
  bool operator()(string s1, string s2) const
  {
    return s1 == s2;
  }
};

struct hash_string
{
  size_t operator()(const string str) const { return hash<const char *>()(str.c_str()); }
};

typedef hash_map<const string, FILE **, hash_string, eqstr> HASH_TYPE;
HASH_TYPE fd_map;

typedef FILE * PFILE;

void usage()
{
  fprintf(stderr, "usage: clfdomainsplit [filenames]\n"
  "This program splits CLF format web logs into multiple files, one for each\n"
  "domain in the input.\n"
  "[-i input] [-d defaultfile] [-c cfg-file] [-o directory]\n"
  "\nVersion: " VERSION "\n");
  exit(ERR_PARAM);
}

FILE **forkit(string &dom);
bool printLine(string &dom, PCCHAR buf);
void closeit();
bool newdom(string &dom);
FILE **openOutput(string &dom);

int main(int argc, char **argv)
{
  for(int i = 0; i < NUM_CHILDREN; i++)
  {
    children[i] = NULL;
  }
  PCCHAR configFile = "/etc/clfdomainsplit.cfg";
  FILE *defFile = stderr;

  int int_c;
  optind = 0;
  while(-1 != (int_c = getopt(argc, argv, "c:d:i:o:")) )
  {
    switch(char(int_c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'c':
        configFile = optarg;
      break;
      case 'd':
        defFile = fopen(optarg, "a");
        if(!defFile)
        {
          fprintf(stderr, "Can't open \"%s\".\n", optarg);
          usage();
        }
      break;
      case 'i':
        input = fopen(optarg, "r");
        if(!input)
        {
          fprintf(stderr, "Can't open \"%s\".\n", optarg);
          usage();
        }
      break;
      case 'o':
        if(chdir(optarg))
        {
          fprintf(stderr, "Can't chdir() to \"%s\".\n", optarg);
          usage();
        }
      break;
    }
  }
  if(parseConfig(configFile))
    return ERR_PARAM;

  defFilePtr = new PFILE;
  *defFilePtr = defFile;

  char buf[4096];
  while(fgets(buf, sizeof(buf), input))
  {
    buf[sizeof(buf) - 1] = '\0';
    strtok(buf, "\r\n");
    string dom;
    if(getDomain(dom, buf))
    {
      fprintf(stderr, "bad line:%s\n", buf);
    }
    else
    {
      if(printLine(dom, buf))
      {
        closeit();
        return 1;
      }
    }
  }
  closeit();
  return 0;
}

bool printLine(string &dom, PCCHAR buf)
{
  if(!fd_map[dom])
  {
    if(newdom(dom))
      return true;
  }
  fprintf(*(fd_map[dom]), "%s\n", buf);
  return false;
}

bool newdom(string &dom)
{
  string canon;
  if(canonDom(canon, dom))
  {
    // send this and anything like it to the default output
    // beware - a file of totally random junk could run us out of RAM!!!
    fd_map[canon] = defFilePtr;
    fd_map[dom] = defFilePtr;
    return false;
  }
  if(fd_map[canon])
  {
    fd_map[dom] = fd_map[canon];
    return false;
  }
  FILE **fpp = openOutput(canon);
  if(!fpp)
    return true;
  fd_map[dom] = fpp;
  return false;
}

FILE **openOutput(string &dom)
{
  if(num_fds > MAX_FDS)
  {
    newTarget++;
    if(newTarget > NUM_CHILDREN)
      newTarget = 0;
  }
  if(newTarget != -1)
  {
    if(children[newTarget] == NULL)
    {
      return forkit(dom); // NB forkit() can call us
    }
    else
    {
      fd_map[dom] = children[newTarget];
      return children[newTarget];
    }
  }

  FILE *fp = fopen(dom.c_str(), "a");
  if(!fp)
  {
    fprintf(stderr, "Can't open/create file \"%s\".\n", dom.c_str());
    return NULL;
  }
  num_fds++;
  FILE **val = new PFILE;
  *val = fp;
  fd_map[dom] = val;
  return val;
}

FILE **forkit(string &dom)
{
  int filedes[2], childFiledes[2];
  if(pipe(filedes) || pipe(childFiledes))
  {
    fprintf(stderr, "Can't create pipe!\n");
    return NULL;
  }
  fflush(NULL);
  int rc = fork();
  if(rc == -1)
  {
    fprintf(stderr, "Can't fork!\n");
    return NULL;
  }
  char c;
  if(rc == 0)
  { // child
    closeit();
    num_fds = 0;
    newTarget = -1;
    num_children = 0;
    for(int i = 0; i < NUM_CHILDREN; i++)
    {
      children[i] = NULL;
    }
    close(childFiledes[0]);
    write(childFiledes[1], &c, 1);
    close(childFiledes[1]);
    close(filedes[1]);
    input = fdopen(filedes[0], "r");
    if(!input)
    {
      fprintf(stderr, "Can't fdopen!\n");
      close(filedes[0]);
      exit(1);
    }
    return openOutput(dom);
  }
  else
  { // parent
    num_children++;
    close(filedes[0]);
    close(childFiledes[1]);
    read(childFiledes[0], &c, 1);
    close(childFiledes[0]);
    FILE *fp = fdopen(filedes[1], "w");
    if(!fp)
    {
      fprintf(stderr, "Can't fdopen!\n");
      return NULL;
    }
    FILE **fpp = new PFILE;
    *fpp = fp;
    children[newTarget] = fpp;
    fd_map[dom] = children[newTarget];
    return children[newTarget];
  }
  return NULL; // should never reach here
}

void closeit()
{
  while(!fd_map.empty())
  {
    HASH_TYPE::iterator st = fd_map.begin();
    FILE **fpp = (*st).second;
    if(fpp)
    {
      if(*fpp)
      {
        fclose(*fpp);
      }
      *fpp = NULL;
    }
    fd_map.erase(st);
  }
}


clfmerge.cpp
#include <stdio.h>
#include <ext/hash_map>
#include <stdlib.h>
#include <map>
#include <cstring>
#include <unistd.h>
#include <ctype.h>

#include "logtools.h"

#define BUF_SIZE 4096

using namespace std;
using namespace __gnu_cxx;

struct eqstr
{
  bool operator()(const char *s1, const char *s2) const
  { return strcmp(s1, s2) == 0; }
};

hash_map<const char *, const char *, hash<const char *>, eqstr> months;

class LogFile
{
public:
  LogFile();
  ~LogFile(){ if(m_fp) fclose(m_fp); }

  // open a file, return 0 for success, 1 for minor error (0 length file)
  // and 2 for serious error (file not found)
  int open(const char *name, bool domain_mangling);

  // get the date string of the current item
  const char *date() const { return m_date; }
  // get the full data for the current item
  const char *line() { m_valid = false; return m_line; }

  // get a line from the file and parse it
  // return 0 for success and 1 for EOF
  int getLine();

  bool valid() const { return m_valid; }

private:
  bool m_valid;
  FILE *m_fp;
  char m_date[17];
  char m_lineBuf[BUF_SIZE];
  char m_lineBuf2[BUF_SIZE + 7];
  char *m_line;
  bool m_web_first;

  // store the date in numeric format so that strcmp() can be used to compare
  // dates.  Returns 1 for error and 0 for OK.
  int setDate();

  LogFile(const LogFile&);
  LogFile & operator=(const LogFile&);
};

LogFile::LogFile()
 : m_valid(false)
 , m_fp(NULL)
{
  m_line = m_lineBuf;
  m_lineBuf[0] = '\0';
}

// Common log format:
// 1.2.3.4 - - [23/Aug/2000:12:00:32 +0200] etc
int LogFile::setDate()
{
  unsigned int i;
  // find '[' or abort if we can't
  for(i = 0; m_lineBuf[i] != '[' && m_lineBuf[i] != '\0'; i++)
  { }

  // if not enough data left for the full date then return
  if(i + 21 > strlen(m_lineBuf))
    return 1;
  memcpy(m_date, &m_lineBuf[i + 8], 4);
  char mon[4];
  memcpy(mon, &m_lineBuf[i + 4], 3);
  mon[3] = '\0';
  const char *m = months[mon];
  if(!m) return 1;
  strcpy(&m_date[4], m);
  memcpy(&m_date[6], &m_lineBuf[i + 1], 2);
  memcpy(&m_date[8], &m_lineBuf[i + 13], 8);
  m_date[16] = '\0';
  if(m_web_first) // make m_lineBuf2 have the data for the mangled line
  {
    unsigned int end_webname;
    // find where the domain name ends
    for(end_webname = 0; m_lineBuf[end_webname] != ' ' && m_lineBuf[end_webname] != '\0'; end_webname++)
    { }

    if(end_webname >= i)
      return 1;
    for(i = 0; i < end_webname; i++)
      m_lineBuf[end_webname] = tolower(m_lineBuf[end_webname]);

    // there will be more than 40 chars in between
    unsigned int start_url = end_webname + 40;
    // search for the start quote character
    for(; m_lineBuf[start_url] != '\"' && m_lineBuf[start_url] != '\0'; start_url++)
    { }
    // search for the space in the web request
    for(; m_lineBuf[start_url] != ' ' && m_lineBuf[start_url] != '\0'; start_url++)
    { }

    if(strlen(&m_lineBuf[start_url]) < 6) return 1;

    memcpy(m_lineBuf2, &m_lineBuf[end_webname + 1], start_url - end_webname);
    m_line = &m_lineBuf2[start_url - end_webname]; // m_line points to next char
    if(strncmp(&m_lineBuf[start_url + 1], "http://", 7))
    {
      strcpy(m_line, "http://");
      m_line += 7;
      memcpy(m_line, m_lineBuf, end_webname);
      m_line += end_webname;
      if(m_lineBuf[start_url + 1] != '/')
      {
        // if URL doesn't start with a '/' then we add one
        *m_line = '/';
        m_line++;
      }
    }
    strcpy(m_line, &m_lineBuf[start_url + 1]);
    m_line = m_lineBuf2;
  }
  return 0;
}

int LogFile::open(const char *name, bool domain_mangling)
{
  m_web_first = domain_mangling;
  m_fp = fopen(name, "r");
  if(!m_fp)
  {
    fprintf(stderr, "Can't open %s.\n", name);
    return 2;
  }
  if(getLine())
    return 1;
  return 0;
}

int LogFile::getLine()
{
  while(1)
  {
    // if can't get more data then return 1
    if(!fgets(m_lineBuf, sizeof(m_lineBuf) - 1, m_fp))
      return 1;
    m_lineBuf[sizeof(m_lineBuf) - 1] = '\0';
    m_line = m_lineBuf;
    strtok(m_line, "\n\r");
    // if setDate() returns 1 then we can't parse the line so we keep looping
    // if setDate() returns 0 then return success!
    if(!setDate())
    {
      m_valid = true;
      return 0;
    }
    fprintf(stderr, "Skipping bad line: %s\n", m_line);
  }
  return 0; // to make compilers happy - will not be reached
}

typedef LogFile *PLogFile;

int item_compare(const void *a, const void *b)
{
  const LogFile * const left = *(LogFile * const *)a;
  const LogFile * const right = *(LogFile * const *)b;
  return strcmp(left->date(), right->date());
}

struct ltstr
{
  bool operator()(const string s1, const string s2) const
  {
    return strcmp(s1.c_str(), s2.c_str()) < 0;
  }
};

void usage(const char *const arg)
{
  fprintf(stderr, "usage: %s [filenames]", arg);
  fprintf(stderr, "\n"
  "This program merges web logs in common log format into a single stream\n"
  "on standard output.  It reads from multiple input files and outputs the\n"
  "data in-order as much as is possible.  If there is only a single input\n"
  "file it will re-order it (with a 1000 line buffer size) to deal with web\n"
  "servers that output data out of order.\n"
  "\nVersion: " VERSION "\n");
  exit(ERR_PARAM);
}

int main(int argc, char **argv)
{
  if(argc == 1)
    return 0;

  unsigned int map_items = 0;
  bool set_map_items = false, domain_mangling = false;
  int int_c;
  optind = 0;
  while(-1 != (int_c = getopt(argc, argv, "b:hd")) )
  {
    switch(char(int_c))
    {
      case '?':
      case ':':
      case 'h':
        usage(argv[0]);
      break;
      case 'b':
        set_map_items = true;
        map_items = atoi(optarg);
      break;
      case 'd':
        domain_mangling = true;
      break;
    }
  }
  months["Jan"] = "01";
  months["Feb"] = "02";
  months["Mar"] = "03";
  months["Apr"] = "04";
  months["May"] = "05";
  months["Jun"] = "06";
  months["Jul"] = "07";
  months["Aug"] = "08";
  months["Sep"] = "09";
  months["Oct"] = "10";
  months["Nov"] = "11";
  months["Dec"] = "12";

  multimap<const string, const string, ltstr> outputMap;

  LogFile **items = new PLogFile[argc - optind];

  unsigned int item_count = 0;
  int i;
  for(i = optind; i < argc; i++)
  {
    items[item_count] = new LogFile;
    int rc = items[item_count]->open(argv[i], domain_mangling);
    // if rc==2 then file not found, if rc==1 then 0 length file
    if(rc > 1)
      return ERR_INPUT;
    if(rc == 1)
      delete items[item_count];
    else
      item_count++;
  }

  if(!set_map_items)
  {
    map_items = item_count * 400;
    if(map_items < 4000)
      map_items = 4000;
  }
  while(item_count > 1)
  {
    qsort(items, item_count, sizeof(LogFile *), item_compare);
    while(items[0]->valid() && strcmp(items[0]->date(), items[1]->date()) <= 0)
    {
      if(map_items > 0)
      {
        outputMap.insert(pair<string, string>(items[0]->date(), items[0]->line()));
        while(outputMap.size() > map_items)
        {
          printf("%s\n", outputMap.begin()->second.c_str());
          outputMap.erase(outputMap.begin());
        }
      }
      else
      {
        printf("%s\n", items[0]->line());
      }
      if(items[0]->getLine())
      {
        delete(items[0]);
        item_count--;
        items[0] = items[item_count];
        break;
      }
    }
  }
  if(item_count == 1)
  {
    if(map_items > 0)
    {
      do
      {
        outputMap.insert(pair<string, string>(items[0]->date(), items[0]->line()));
      } while(!items[0]->getLine() && outputMap.size() < map_items);

      if(items[0]->valid())
      {
        do
        {
          outputMap.insert(pair<string, string>(items[0]->date(), items[0]->line()));
          CPCCHAR tmp = outputMap.begin()->second.c_str();
          if(printf("%s\n", tmp) != int(strlen(tmp) + 1))
          {
            fprintf(stderr, "Can't write output!\n");
              return ERR_OUTPUT;
          }
          outputMap.erase(outputMap.begin());
        } while(!items[0]->getLine());
      }
      delete items[0];
      while(!outputMap.empty())
      {
        CPCCHAR tmp = outputMap.begin()->second.c_str();
        if(printf("%s\n", tmp) != int(strlen(tmp) + 1))
        {
          fprintf(stderr, "Can't write output!\n");
            return ERR_OUTPUT;
        }
        outputMap.erase(outputMap.begin());
      }
    }
    else
    {
      do
      {
        CPCCHAR tmp = items[0]->line();
        if(printf("%s\n", tmp) != int(strlen(tmp) + 1))
        {
          fprintf(stderr, "Can't write output!\n");
            return ERR_OUTPUT;
        }
      } while(!items[0]->getLine());
    }
  }
  delete items;
  return 0;
}
